query/src/main/scala/filodb/query/exec/SelectRawPartitionsExec.scala

```diff
@@ -22,6 +23,29 @@ object SelectRawPartitionsExec {
             schema.data.columns.find(c => c.name == "max"  && c.columnType == DoubleColumn).map(_.id)
           }
   }
+
+  def findFirstRangeFunction(transformers: Seq[RangeVectorTransformer]): Option[RangeFunctionId] =
+    transformers.collect { case p: PeriodicSamplesMapper => p.functionId }.headOption.flatten
+
+  def replaceRangeFunction(transformers: Seq[RangeVectorTransformer],
+                           oldFunc: Option[RangeFunctionId],
+                           newFunc: Option[RangeFunctionId]): Seq[RangeVectorTransformer] =
+    transformers.map {
+      case p: PeriodicSamplesMapper if p.functionId == oldFunc => p.copy(functionId = newFunc)
+      case other: RangeVectorTransformer => other
+    }
+
+  // Optimize transformers, replacing range functions for downsampled gauge queries if needed
+  def optimizeForDownsample(schema: Schema, transformers: Seq[RangeVectorTransformer]): Seq[RangeVectorTransformer] = {
```

Comment:
The dataset has many schemas nd should not have `hasDownsampledData`.  It may even mix up downsampled and normal schemas.

"generating wrong plan"... what is meant by this?  WE really have just two choices:
- generate a generic, schema-agnostic plan at planning time, then generating a more specific one later (current approach chosen)
OR - generate a schema-specific, final plan at planning time.